dimod/response/response.py

```diff
@@ -0,0 +1,439 @@
+"""TODO:
+    - module level docstring
+    - examples in Response
+"""
+from collections import namedtuple
+
+import pandas as pd
+
+from dimod.compatibility23 import iteritems
+from dimod.vartypes import Vartype
+
+try:
+    import dwave_micro_client as microclient
+except ImportError:  # pragma: no cover
+    pass
+
+__all__ = ['Response']
+
+
+class Response(object):
+    """Encodes a response from a dimod sampler.
+
+    Args:
+        vartype (:class:`.Vartype`/str/set):
+            The variable type desired for the response. Accepted input values:
+            :class:`.Vartype.SPIN`, ``'SPIN'``, ``{-1, 1}``
+            :class:`.Vartype.BINARY`, ``'BINARY'``, ``{0, 1}``
+
+    Attributes:
+        vartype (:class:`.Vartype`): The variable type of the response.
+
+    """
+
+    def __init__(self, vartype):
+        # the response object keeps two dataframes, these are kept index-linked so that they can
+        # be joined
+        self.df_samples = pd.DataFrame(dtype='int8')
+        self.df_data = pd.DataFrame(columns=['energy'])
+
+        try:    # pragma: no cover
+            if isinstance(vartype, str):
+                vartype = Vartype[vartype]
+            else:
+                vartype = Vartype(vartype)
+
+            if not (vartype is Vartype.SPIN or vartype is Vartype.BINARY):
+                raise ValueError  # this gets caught
+        except (ValueError, KeyError):    # pragma: no cover
+            raise TypeError(("expected input vartype to be one of: "
+                             "Vartype.SPIN, 'SPIN', {-1, 1}, "
+                             "Vartype.BINARY, 'BINARY', or {0, 1}."))
+        self.vartype = vartype
+
+        # we also store a list of futures that are awaiting read
+        self._futures = []
+
+    def __len__(self):
+        """The number of samples."""
+        return self.df_samples.shape[0]
+
+    def __iter__(self):
+        """Iterate over the samples as dicts from low energy to high."""
+        return self.samples()
+
+    def __str__(self):
+        return self.merged(sorted_by_energy=False).__str__()  # pragma: no cover
+
+    @property
+    def df_samples(self):
+        """:class:`pandas.DataFrame`: The samples. Should be treated as read-only."""
+        # if there are any waiting futures, process them now
+        if self._futures:
+            self._add_samples_future()
+
+        return self._samples
+
+    @df_samples.setter
+    def df_samples(self, df_samples):
+        self._samples = df_samples
+
+    @property
+    def df_data(self):
+        """:class:`pandas.DataFrame`: The data. Should be treated as read-only."""
+        # if there are any waiting futures, process them now
+        if self._futures:
+            self._add_samples_future()
+
+        return self._data
+
+    @df_data.setter
+    def df_data(self, df_data):
+        self._data = df_data
+
+    def merged(self, sorted_by_energy=True):
```

Comment:
Nice!

Reply:
Though I did find a case where this breaks... if there is a variable named `'energy'` as well as one named `'energy_data'` this will result in two columns named `'energy_data'`; worse if there is only one named `'energy_data'` it will result in the energies being stored under `'energy'` while another column is called `'energy_data'` (no name conflict, but opposite behaviour of the case where the energies are stored under `'energy_data'` because another column was named `'energy'`)